azure/converters/image.go

```diff
@@ -69,8 +69,8 @@ func computeImageToSDK(image *infrav1.Image) (*compute.ImageReference, error) {
 	if image.ComputeGallery.ResourceGroup != nil && image.ComputeGallery.SubscriptionID != nil {
 		return &compute.ImageReference{
 			ID: to.StringPtr(fmt.Sprintf(idTemplate,
-				image.ComputeGallery.SubscriptionID,
-				image.ComputeGallery.ResourceGroup,
+				to.String(image.ComputeGallery.SubscriptionID),
+				to.String(image.ComputeGallery.ResourceGroup),
```

Comment:
This function must return a _valid_ `ImageReference`. Since these two fields are optional, I would hope that a valid `ImageReference` can omit either of them, but I'm not sure what it would look like. Since `to.String` returns an empty string when its input is nil, the resulting `ImageReference` would look something like this if both fields are nil:

`/subscriptions//resourceGroups//providers/Microsoft.Compute/galleries/FAKEGALLERYID/images/FAKEIMAGEID/versions/FAKEVERSION`

Is wonder if there's a some Azure library that can validate an `ImageReference`. It would be nice to call it in the unit test.

Reply:
The `ID` is an opaque field that isn't checked until runtime, but I'm fairly certain that omitting either subscription or resourceID from it makes it invalid (hence the `nil` check guarding this code). I didn't see any specific validation func in the SDK, but you could always call GET on the resource.

Reply:
> (hence the nil check guarding this code)

Thanks for this. I should read the comment on L67.

Reply:
On a related note, is it possible for this function to return an invalid output? For example, is it possible that `image.SharedGallery`, `image.ComputeGallery` are both nil? (It's definitely possible in the code, but perhaps impossible thanks to a validation webhook on the corresponding Kubernetes resource).

Reply:
The API function `ImageToSDK` validates first if either of `image.SharedGallery` or `image.ComputeGallery` are not `nil` before calling this function. https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/converters/image.go#L36. but I agree if this function is called without such prior validation here or in webhook, it will fail.

Reply:
Maybe this function could be more defensive and return an error if either of the non-`nil` strings is empty, and we can cover this case in the unit test. @supershal do you think that's worth doing or is the webhook adequate?

Reply:
+1. It would be better to introduce check here. I will make the suggested fix.

Reply:
> Maybe this function could be more defensive and return an error if either of the non-nil strings is empty, and we can cover this case in the unit test.

(I think the webhook is responsible for validating the _contents_ of the strings.)

Reply:
+1 to a defensive check that at least one of `image.ComputeGallery` or `image.SharedGallery` is not nil.

Reply:
Thank you. added `nil` check.